CI tests for ASAN, `make install` and `PERL_UNICODE`

nwc10 opened the pull request:
This branch adds several new CI tests, refactors `testsuite.yml` somewhat, but probably **reduces** the wallclock time for all CI tests to complete.

The "critical path" for all tests to complete is

1. Sanity (23m)
2. cygwin (52m)

This branch

1. replaces `make test` with `make test_harness` in Sanity (`test` does not run tests in parallel) - saving at least 6m
2. replace `make` and `make test` with `make test_prep` and `./perl harness` - saving roughly 1m

`make -j2 test` first needs to (re)check that all the extensions are up to date before getting to the real work - on Cygwin this takes at least 40s.

This likely reduces the total wallclock time to complete all tests by about 10 minutes. The elapsed time seems quite variable, I assume due to load from other VMs on the same host. cygwin is still the slowest test - the addition of the ASAN tests does not change the critical path to completion.

New tests

1. ASAN (so `malloc`/`free` goofs, but also agressive leak checking)
2. `make install`
3. `PERL_UNICODE=""` and UTF-8 locales
4. shared libraries on macOS
5. minitest
5. building **without** `-DDEBUGGING`
6. Adding `MALLOC_PERTURB_` and `MALLOC_CHECK_` to (most) Linux builds

We already had configurations that looked like they were testing `PERL_UNICODE=""` and UTF-8 locales, but these didn't actually do anything special.

This branch incorporates #19115, as without those fixes the ASAN builds fail.

nwc10 committed:
Avoid leaking a scalar body after REGEXP to PVLV assignment

PVLV scalars are used to implement various obscure features of perl, such as
elements in tied hashes and hash lookups used as function parameters. To
make these work, PVLVs must be capable of holding all scalar types. First
class REGEXPs are larger than PVLVs, hence there is special case code to
attach REGEXP bodies to PVLVs, for code that both dereferences a regex object
and then assigns that value to a PVLV scalar. This was first implemented in
Oct 2012 by commit 8d919b0a35f2b57a:
    Allow regexp-to-pvlv assignment

and improved in July 2017 with commit df6b4bd56551f2d3:

It turns out that the implementation would leak a scalar body, for the
"normal" use case (of this obscure feature). This wasn't noticed for two
reasons

1) bodies are allocated from arenas, and arenas are correctly freed during
   full destruction.
2) the body is not leaked if the the PVLV scalar is then modified

The regression tests added by commit 8d919b0a35f2b57a were also testing
the corner case of concatenating to one of these values - ie that
sv_force_normal_flags() handled unwinding the special cases correctly.
To avoid code duplication, the tests added by that commit caused all PVLV
scalars to be passed to sv_force_normal_flags(), so didn't actually test
regular destruction of such scalars.

The tests added in Aug 2017 by commit 622c613e12cef84c:
    PVLV-as-REGEXP: avoid PVX double free

    With v5.27.2-30-gdf6b4bd, I changed the way that PVLVs store a regexp
    value (by making the xpv_len field point to a regexp struct). There was a
    bug in this, which caused the PVX buffer to be double freed.

    Several REGEXP SVs can share a PVX buffer. Only one of them will have a
    non-zero xpv_len field, and that SV gets to free the buffer.

    After the commit above, the non-zero xpv_len was triggering an extra free.

    This was showing up in smokes as failures in re/recompile.t when invoked
    with PERL_DESTRUCT_LEVEL=2 (which t/TEST does).

were actually the first to expose this bug, even though it had been present
since 2012.

The bug is only visible with a full leak test (eg valgrind --leak-check=full
or an ASAN build) with -DPURIFY (to disable arenas), hence why it hasn't
been noticed.

nwc10 committed:
Don't leak in hv_common when croaking about PL_strtab

hv_common can perform read-only actions on PL_strtab, but not write actions.
The code that detects this and croaks had been just after the allocation of
a new HE *, and hence was leaking it. Re-order the code to avoid the leak.

The leak usually wasn't noticeable as HEs are allocated from arenas, and
arenas are correctly freed during full destruction. However, building with
-DPURIFY replaces arenas with individual allocations, making this leak
visible. It's unlikely to have been hit by any production code, but it was
causing leaks during the regression tests.

Also change embed.fnc so that S_new_HE's prototype is not declared under
-DPURIFY, as the static function itself is not defined in this case. This
fixes a compiler warning.

nwc10 committed:
Add "missing" descriptions to testsuite.yml

If it's worth doing, it's worth overdoing.

nwc10 committed:
Add a CI job to confirm that minitest keeps passing

Currently minitest passes - let's keep it that way.

minitest might feel like makework, until one makes a seemingly innocent
change that breaks the build in "inconceivable" ways and it's no longer
even possible to get to the point of being able to run `make test`.

nwc10 committed:
A prototype CI job to build with Address Sanitizer

Whilst this *will* detect various dangerous mistakes involving bad memory
accesses, it's far more likely to spot if changes create memory leaks.

nwc10 committed:
All workflow jobs use the same env: so move it to the top level

Also nothing references $ENV{WORKSPACE} so no need to set this. Documentation
for actions/checkout implies that it is reading from $GITHUB_WORKSPACE.

nwc10 committed:
Workflows should `make test_harness` to get parallel testing

`make test` uses t/TEST which always runs tests sequentially.
TEST_JOBS only makes sense with t/harness, which is run by the
`test_harness` target.

nwc10 committed:
Test ASAN builds with and without -DPURIFY

Also we can fail-fast on ASAN to save some resources. Likely if one ASAN
build fails, the others won't tell us about different problems.

nwc10 committed:
In the Linux workflows use MALLOC_PERTURB_ and MALLOC_CHECK_

MALLOC_CHECK_=3 causes glibc to abort if it detects malloc errors.
Specifically, 3 will be "Print detailed error message, stack trace, and
memory mappings, and abort the program."

MALLOC_PERTURB_=254 will cause newly allocated memory to be set to non-zero
values, and freed memory to be set to the given non-zero value.

See `man mallopt`

Intentionally only setting these for the build and test steps, but not the
Configure step, to see if we can later expose bugs caused by bad probes.
If we set it for Configure, the abort this triggers might simply cause a
probe to be silently failed, and we could miss problems.

nwc10 committed:
CI tests for PERL_UNICODE="" with a UTF-8 locale

nwc10 committed:
PERL_UNICODE= and LANG= can't be set as Configure arguments

The intent is to test with non-C locales and non-default Unicode handling,
but to do this the values need to be set as environment variables for the
test (and build - occasionally bugs are not exposed by the testsuite).

Configure happily accepts anything on the command line, and records all its
command line arguments in config.sh (and hence also within %Config::Config).
But nothing acts on them.

It's easier to set these in a separate workflow job (and test only a couple)
rather than creating a second environment matrix in this job, and hence
test $m * $n combinations.

Unfortunately the GH test machines only have the locales C, C.UTF-8, POSIX
and en_US.utf8, so we can't easily test anything more "interesting" than
en_US.utf8.

nwc10 committed:
The CI workflow needs some test builds without -DDEBUGGING

nwc10 committed:
Test `make install` in the GH workflow

nwc10 committed:
Run `./perl t/harness` directly in the CI workflow

`make -j2 test_harness` causes make to re-run the build in all the extension
directories, which can take 20 seconds even on fast server hardware. Hence
change the build target to `test_prep`, the Makefile pre-requisite for
testing, and then run t/harness directly with the same environment as the
Makefile and ./runtests would have set up. As we already set
PERL_SKIP_TTY_TEST=1 in the global environment we don't have anything else
to set up.

Keep the 'linux-i386' and 'install' jobs on the regular `test` and
`test_harness` targets to ensure we still cover them.

nwc10 committed:
Test clang's ASAN as well as gcc's ASAN

On the GH virtual machine, gcc's ASAN appears to be slightly buggy, as the
`-Accflags=-DPURIFY` job is causing cpan/File-Path/t/Path.t to reliably
SEGV after passing all 167 subtests:
    Tracer caught signal 11: addr=0x0 pc=0x7ff95dfaa8b0 sp=0x7ff95a2a6ce0
    LeakSanitizer has encountered a fatal error.

I can't replicate this on machines I have access to, which makes it rather
hard to debug. Hence side-step the issue by running this test with clang
instead, so that we also test with clang's ASAN implementation.

We still test with gcc's ASAN in one of the PERL_UNICODE jobs.

nwc10 committed:
Test a macOS threaded shared-library build too

nwc10 commented:
And now cygwin hates me, but for some reason blead is fine.

hvds commented on sv.c:
```diff
@@ -6788,8 +6789,10 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
                  * regex body. The PVX buffer is shared by multiple re's
                  * and only freed once, by the re whose len in non-null */
-                STRLEN len = ReANY(sv)->xpv_len;
+                struct regexp *r = ((XPV*)SvANY(sv))->xpv_len_u.xpvlenu_rx;
```
Unless I've missed a reason, I think this should stick with `struct regexp *r = ReANY(sv);` - I see no benefit to manually expanding the inline function, and some risk if `ReANY` changes again in the future.

Also in the commit message,  I think you intended to follow `with commit df6b4bd56551f2d3:` with that commit's subject line `give REGEXP SVs the POK flag again`.

nwc10 replied:
> Unless I've missed a reason, I think this should stick with `struct regexp *r = ReANY(sv);` - I see no benefit to manually expanding the inline function, and some risk if `ReANY` changes again in the future.

The diff hunk doesn't really give enough context. What happened was..

I started with code in `sv.c` that reads like this (in blead)

```
            if (isREGEXP(sv)) {
                /* SvLEN points to a regex body. Free the body, then
                 * set SvLEN to whatever value was in the now-freed
                 * regex body. The PVX buffer is shared by multiple re's
                 * and only freed once, by the re whose len in non-null */
                STRLEN len = ReANY(sv)->xpv_len;
                pregfree2((REGEXP*) sv);
                SvLEN_set((sv), len);
                goto freescalar;
            }
```

and I read it, and thought "wait, what's going on here. The comment mentions `SvLEN` but I don't see any sign of `SvLEN`."

So then I had to figure out that the `ReANY` is actually what the `SvLEN` is talking about in the comment - it's not obvious

And then I had to go look up the definition of that

```
PERL_STATIC_INLINE struct regexp *
Perl_ReANY(const REGEXP * const re)
{
    XPV* const p = (XPV*)SvANY(re);

    PERL_ARGS_ASSERT_REANY;
    assert(isREGEXP(re));

    return SvTYPE(re) == SVt_PVLV ? p->xpv_len_u.xpvlenu_rx
                                   : (struct regexp *)p;
}
```

at which point I thought "eh, what's that ternary doing? Sometimes it isn't `SvLEN`"

So then I had to re-read the code in `sv.c`.

Enough to look for more context:

```
        case SVt_PVLV:
            if (LvTYPE(sv) == 'T') { /* for tie: return HE to pool */
                SvREFCNT_dec(HeKEY_sv((HE*)LvTARG(sv)));
                HeNEXT((HE*)LvTARG(sv)) = PL_hv_fetch_ent_mh;
                PL_hv_fetch_ent_mh = (HE*)LvTARG(sv);
            }
            else if (LvTYPE(sv) != 't') /* unless tie: unrefcnted fake SV**  */
                SvREFCNT_dec(LvTARG(sv));
            if (isREGEXP(sv)) {
                /* SvLEN points to a regex body. Free the body, then
                 * set SvLEN to whatever value was in the now-freed
                 * regex body. The PVX buffer is shared by multiple re's
                 * and only freed once, by the re whose len in non-null */
                STRLEN len = ReANY(sv)->xpv_len;
                pregfree2((REGEXP*) sv);
                SvLEN_set((sv), len);
                goto freescalar;
            }
```

and then I had enough to mentally simplify the code in `Perl_ReANY` - it was correct, and it always took one side of the ternary.

And that was a bunch of work, involving two context switches. I didn't find this very readable.

Hence why I deliberately decided to inlinine - based on this experience I felt that it made the code more readable:

```
            if (isREGEXP(sv)) {
                /* SvLEN points to a regex body. Free the body, then
                 * set SvLEN to whatever value was in the now-freed
                 * regex body. The PVX buffer is shared by multiple re's
                 * and only freed once, by the re whose len in non-null */
                struct regexp *r = ((XPV*)SvANY(sv))->xpv_len_u.xpvlenu_rx;
                STRLEN len = r->xpv_len;
                pregfree2((REGEXP*) sv);
                del_body_by_type(r, SVt_REGEXP);
                SvLEN_set((sv), len);
                goto freescalar;
            }
 ```


I also don't think that there's real risk here - this code is "inside" the abstraction of how regexs are store in two possible places, given that it's responsible for correctly freeing up one of those two possible places. If `ReANY` needs to change, then this code needs to be changed too, or at least audited that it's still correct.


So that's how I saw it. It didn't feel like using `ReANY` worked as an abstraction, because it didn't seem to be something you could ignore while trying to understand (or verify) the code - it was getting in the way, rather than making it simpler.

hvds replied:
> If `ReANY` needs to change, then this code needs to be changed too, or at least audited that it's still correct.

In such circumstance, I'd have imagined one would primarily rely on grepping for ReANY to find what needs auditing. I'd suggest at least mentioning it in the comment: `SvLEN points to a regex body (the PVLV case in ReANY).`.

(More formally, I think it dubious to assert that SvLEN does the pointing, but I guess that's a reasonable enough shorthand.)

nwc10 replied:
> I'd suggest at least mentioning it in the comment: `SvLEN points to a regex body (the PVLV case in ReANY).`.

I started tweaking the comments. And then tried to remember exactly what was happening where and when...

It was only a week ago, and I'd already started to forget all the complexity (possibly too clever for its own good, but I can't spot a cleaner way...)

So I've just pushed commit 07e0407ab51f4df8:

```
    Add comments describing how PVLVs store REGEXPs by reference

 inline.h  | 15 +++++++++++++++
 regcomp.c | 10 ++++++++++
 sv.c      | 27 +++++++++++++++++++++++----
 3 files changed, 48 insertions(+), 4 deletions(-)
```

Quite a lot of comments. Do these work better than what we had?

If you think that they are good enough to merge, as part of the rebase I was proposing to re-order the commits so that this commit with comments was immediately after the 2 commits with code changes, and before the commits that only deal with `.github/workflows/testsuite.yml`

hvds replied:
> Quite a lot of comments. Do these work better than what we had?

I think so, yes.

> If you think that they are good enough to merge, as part of the rebase I was proposing to re-order the commits so that this commit with comments was immediately after the 2 commits with code changes, and before the commits that only deal with `.github/workflows/testsuite.yml`

That makes sense to me.

hvds replied:
> If you think that they are good enough to merge

I don't see any blocking issues, but if there's a chance to get eyeballs from someone that knows either of these areas of code better then that'd be a help. If that's unlikely any time soon, then pushing it to see what breaks also works (at least if we're comfortable on the github accounting policy).

nwc10 replied:
We don't seem to be getting more eyeballs.

Even if we punt on the workflow changes until we get a better answer on "quota", I'd quite like to get the 3 code commits into blead, because they make the testsuite 100% clean under ASAN (ie no leaks). Right now every time I've done an ASAN build on blead or anything branched from it, I've had to re-confirm that the 3 tests I see failing are the 3 that I know "should" fail. Getting blead clean (and keeping it that way) would be good.

hvds replied:
> We don't seem to be getting more eyeballs.
> 
> Even if we punt on the workflow changes until we get a better answer on "quota", I'd quite like to get the 3 code commits into blead

I'd go for it, we're early enough in the release cycle to have lots of time.

hvds commented on .github/workflows/testsuite.yml:
```diff
@@ -464,3 +464,53 @@ jobs:
+        # exercise a variety of build options
+        # threads often cause build issues
+        CONFIGURE_ARGS:
+          - "-Dusethreads"
```
Do we want/need also with and without debugging for this ASAN build? I've been noticing recently that gcc can act quite differently (and sometimes buggily) in the presence of asserts. I've no clue what the cost/benefit ratio looks like though.

nwc10 replied:
> Do we want/need also with and without debugging for this ASAN build? I've been noticing recently that gcc can act quite differently (and sometimes buggily) in the presence of asserts. I've no clue what the cost/benefit ratio looks like though.

I don't either. In that, I believe that everything runs in parallel, and we don't seem to saturate the available hardware, meaning that "however many" Linux tests we add, cygwin will still finish last. Between the "ASAN" and "PERL_UNICODE" there are two jobs with `-Dusethreads` and only one with (implied) `-Uusethreads` so I guess moving the precise location of the `-DDEBUGGING` and having one (more) job here with (`no threads` and `no DEBUGGING`) would cover this case.

@rjbs - do you know what accounting policy github has for workflow jobs, and if we're at risk of hitting it and hence having useful jobs denied?

hvds replied:
> @rjbs - do you know what accounting policy github has for workflow jobs, and if we're at risk of hitting it and hence having useful jobs denied?

I recall we tripped up on that not so long ago, but I don't remember the details; I think it may have been @atoomic that fixed things that time.

nwc10 replied:
> > @rjbs - do you know what accounting policy github has for workflow jobs, and if we're at risk of hitting it and hence having useful jobs denied?
> 
> I recall we tripped up on that not so long ago, but I don't remember the details; I think it may have been @atoomic that fixed things that time.

I can't find *much* documentation that gives me confidence in an answer.


The only thing that mentions limits (well, free quotas) is https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions

I assume that we are "GitHub Free for organizations" and hence that gives us 2,000 minutes per month.

Which works out at 64 minutes daily. Which I think we'd burn through on one push request.

Which must mean that this applies:

> GitHub Actions usage is free for both public repositories and self-hosted runners.

Which seems to imply that it's unlimited. Or more realistically, "unlimited".

Sigh. So I still don't *really* know.

hvds replied:
> > GitHub Actions usage is free for both public repositories and self-hosted runners.
> 
> Which seems to imply that it's unlimited. Or more realistically, "unlimited".

Ah, my recollection of tripping up on this before must have been for the security repo, which is _not_ a public repository. I think we had the mirroring set up such that every CI run for the public repo was duplicated; and that we eventually worked out what was happening and changed the set up so that the duplication no longer occurs.

So I think you're right, that we are "unlimited".

hvds commented on .github/workflows/testsuite.yml:
```diff
@@ -53,6 +53,10 @@ jobs:
 
+  # "sanity_check" tests -Dusethreads without -DDEBUGGING
+  # "linux-i386" tests -Uusethreads without -DDEBUGGING
+  # All other builds test with -DDEBUGGING
```
As far as I can see, it's just `linux`, `make install` and `cygwin` that build with DEBUGGING. Was this intended to say "all other _linux_ builds"?

nwc10 replied:
Good catch. Fixed in the (partial) rebase.

Thanks

hvds commented on .github/workflows/testsuite.yml:
```diff
@@ -145,14 +151,14 @@ jobs:
       - name: Run Tests
         run: |
-          MALLOC_PERTURB_=254 MALLOC_CHECK_=3 TEST_JOBS=2 make -j2 test_harness
+          LD_LIBRARY_PATH=`pwd` MALLOC_PERTURB_=254 MALLOC_CHECK_=3 TEST_JOBS=2 ./perl t/harness
```
Why "`pwd`" here rather than "." as above? Should the above be changed too? (And similarly for the FIXUP commit.)

nwc10 replied:
"Because it changes directory, so the lazy approach used previously doesn't work"...

Yes, good catch. I added a commit in the rebase that changes all those `.`s to ```pwd```s

Thanks.

hvds reviewed (commented):
This changed to draft status while I was looking at it, but I'll dump some comments here anyway.

nwc10 committed:
Run t/TEST directly on cygwin in the CI workflow

`make -j2 test` causes make to re-run the build in all the extension
directories, which takes about 40 seconds on the GH servers - ie this alone
is roughly 1% of the total test time.

Hence change the build target to `test_prep`, the Makefile pre-requisite for
testing, and then run t/TEST.pl directly with the same environment as the
Makefile and ./runtests would have set up. As we already set
PERL_SKIP_TTY_TEST=1 in the global environment we don't have anything else
to set up.

nwc10 committed:
Change all LD_LIBRARY_PATH setup from '.' to `pwd`

The previous commit added various new lines that set LD_LIBRARY_PATH=`pwd`,
whereas the existing lines that set LD_LIBRARY_PATH set it to '.'.

LD_LIBRARY_PATH needs to be set to the build directory to run an uninstalled
perl binary *if* that binary is built with a shared perl library. It's not
needed for the default configuration, but to simplify the test definitions we
were doing it always.

In turn, LD_LIBRARY_PATH doesn't need to be set to an *absolute* path if all
that is invoked is the perl binary - it doesn't change directory, and it
does shell out to itself, hence setting LD_LIBRARY_PATH=. is just fine here.

Running the regression tests *does* involve changing directory before
invoking perl, so LD_LIBRARY_PATH needs to be set to an absolute path there.
Running `./perl -Ilib -V` doesn't change directory (or shell out) so '.'
was good enough.

There's no real harm in also invoking `pwd` to get an absolute path here, so
do this instead of just coding '.', as the consistency makes the workflow
easier to read - you don't get distracted wondering "Why are these
different, and is that a bug?"

nwc10 commented:
> This changed to draft status while I was looking at it, but I'll dump some comments here anyway.

I changed it to draft because I pushed it thinking that it was all going to be good, and I only realised later that to my horror I'd done something that broke cygwin. It was only the YAML workflow that was really draft - the first two commits are in the existing PR that I referenced.

And then I found that my timing numbers were off.

I think that all that is now fixed.

nwc10 committed:
Test ASAN with -DDEBUGGING (and so also with assertions enabled)

Make it clear in the earlier comment about -DDEBUGGING that it was
referencing the "linux" job, not the entire file.

nwc10 committed:
Add comments describing how PVLVs store REGEXPs by reference

hvds commented on sv.c:
```diff
@@ -6785,10 +6788,26 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
+                 * We get the body pointer directly from the union, free it,
+                 * then set SvLEN to whatever value was in the now-freed regex
+                 * body. The PVX buffer is shared by multiple re's and only
+                 * freed once, by the re whose SvLEN in non-null.
```
s/in/is/

nwc10 replied:
Pants!

(Well spotted)

Do I really rebase again again to fix this? Or do I just try to remember to fix this at the point that we merge?

hvds replied:
My approach would be to push a fixup commit, and fix it up via rebase just before merging. But that's based on my own workflows - I know I'd always look again at the commits before merging, so I'd spot the "fixup" message and act on it. If you're not confident you'd spot that, it's fine to push it as a normal commit, or leave it out - it's no great harm if we fix it with a normal commit after the fact, just another bit of churn in the logs.

nwc10 replied:
I fixed it before testing the rebase.

jkeenan commented:
On 9/23/21 10:47 AM, Hugo van der Sanden wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In sv.c <https://github.com/Perl/perl5/pull/19140#discussion_r714873816>:
> 
>> @@ -6788,8 +6789,10 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
>                    * set SvLEN to whatever value was in the now-freed
>                    * regex body. The PVX buffer is shared by multiple re's
>                    * and only freed once, by the re whose len in non-null */
> -                STRLEN len = ReANY(sv)->xpv_len;
> +                struct regexp *r = ((XPV*)SvANY(sv))->xpv_len_u.xpvlenu_rx;
> 
>     If you think that they are good enough to merge
> 
> I don't see any blocking issues, but if there's a chance to get eyeballs 
> from someone that knows either of these areas of code better then that'd 
> be a help. If that's unlikely any time soon, then pushing it to see what 
> breaks also works (at least if we're comfortable on the github 
> accounting policy).
> 

Please consider a smoke-me branch as well, so that we can pick up 
additional OSes.

nwc10 commented:
> Please consider a smoke-me branch as well, so that we can pick up additional OSes.

I took the branch that this PR is coupled to, and rebased the commits onto blead as `smoke-me/nicholas/ASAN-PURIFY-github` two days ago. I haven't spotted any problems. (Am I missing them?)

nwc10 commented:
We have fails *and* passes for cygwin and i386 (for the same commit)

```
fatal: reference is not a tree: 07e0407ab51f4df89b633c7f5c7babb226161562
```

I think that this is because I

1. accidentally pushed my branch to `origin` (ie this repository)
2. then realised my mistake and deleted it
3. then pushed the updated branch to `nwc10` (where it belongs)

and the CI run was triggered by the first push, for the given SHA1, but at the point that the CI run for the first push started on those two hosts, that SHA1 had been deleted again (but the run wasn't cancelled) hence this error.

ie it's not a problem with the workflow. (which matters, when most of the branch is about workflows)
